File: OI-ACM/Math/NumberTheory/Extgcd/minmax.py
import matplotlib.pyplot as plt
import math


def extgcd(a, b):
    if b == 0:
        return (1, 0, a)
    else:
        x, y, gcd = extgcd(b, a % b)
        return (y, x - a // b * y, gcd)


def minmax(a, b, c, n):
    s, t, gcd = extgcd(a, b)
    print(f'{a}x+{b}y={c}')
    print(f'(s, t) = ({s}, {t})')
    if c < 0:
        return -1
    elif c == 0 and gcd == 0:
        return (0, n)
    elif c == 0:
        return (0, 0)
    elif gcd == 0 or c % gcd != 0:
        return -1
    x0 = s * (c // gcd)
    y0 = t * (c // gcd)
    print(f'(x0, y0) = ({x0}, {y0})')
    k_min = math.ceil(-x0 / (b // gcd))
    k_max = math.floor(y0 / (a // gcd))
    if k_min > k_max:
        return -1
    print(f'k in [{k_min, k_max}]')
    x_min = 1e8
    x_max = -1
    x = []
    y = []
    for k in range(k_min, k_max + 1):
        x_k = x0 + k * (b // gcd)
        y_k = y0 - k * (a // gcd)
        if x_k + y_k <= n:
            x.append(x_k)
            y.append(y_k)
            x_min = min(x_min, x_k)
            x_max = max(x_max, x_k)
    print(f'x in [{x_min, x_max}]')
    plt.plot(x, y, '-o')
    plt.show()
    return (x_min, x_max)


minmax(2, 1, 10, 10)

File: OI-ACM/Math/NumberTheory/Extgcd/minmax.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

ll extgcd(ll a, ll b, ll &x, ll &y) {
    ll d;
    if (b == 0)
        x = 1, y = 0, d = a;
    else {
        d = extgcd(b, a % b, y, x); // y = x',  x = y'
        y -= a / b * x;             // y = y - a / b * x = x' - a / b * y'
    }
    return d;
}

void minmax(ll a, ll b, ll c, ll n) {
    ll x0, y0, d = extgcd(a, b, x0, y0);
    if (c < 0) {
        cout << -1 << '\n';
        return;
    } else if (c == 0 && d == 0) {
        cout << 0 << ' ' << n << '\n';
        return;
    } else if (c == 0) {
        cout << 0 << ' ' << 0 << '\n';
        return;
    } else if (d == 0 || c % d != 0) {
        cout << -1 << '\n';
        return;
    }
    // c > 0, gcd != 0, c % gcd == 0
    x0 *= c / d;
    y0 *= c / d;
    ll p = b / d, q = a / d;
    ll k_min = ceil((0.0 - x0) / p);
    ll k_max = floor((y0 - 0.0) / q);
    if (k_min > k_max) {
        cout << -1 << '\n';
        return;
    }
    ll x_min = 1e8, x_max = -1;
    for (ll k = k_min; k <= k_max; k++) {
        ll x_k = x0 + k * p;
        ll y_k = y0 - k * q;
        if (x_k + y_k <= n) {
            x_min = min(x_min, x_k);
            x_max = max(x_max, x_k);
        }
    }
    cout << x_min << ' ' << x_max << '\n';
}
signed main() {
    ll p, q, n1, n2, n3;
    cin >> p >> q >> n1 >> n2 >> n3;
    minmax(n3 - n1, n2 - n1, q - n1 * p, p);
}